src/VirtoCommerce.CatalogModule.Data/Services/DynamicAssociationService.cs

```diff
@@ -15,76 +19,114 @@ namespace VirtoCommerce.CatalogModule.Data.Services
 {
     public class DynamicAssociationService : IDynamicAssociationService
     {
-        private readonly Func<ICatalogRepository> _repositoryFactory;
+        private readonly Func<ICatalogRepository> _catalogRepositoryFactory;
+        private readonly IPlatformMemoryCache _platformMemoryCache;
+        private readonly IEventPublisher _eventPublisher;
 
-        public DynamicAssociationService(Func<ICatalogRepository> repositoryFactory)
+        public DynamicAssociationService(
+            Func<ICatalogRepository> catalogRepositoryFactory,
+            IPlatformMemoryCache platformMemoryCache,
+            IEventPublisher eventPublisher)
         {
-            _repositoryFactory = repositoryFactory;
+            _catalogRepositoryFactory = catalogRepositoryFactory;
+            _platformMemoryCache = platformMemoryCache;
+            _eventPublisher = eventPublisher;
         }
 
         public async Task<DynamicAssociation[]> GetByIdsAsync(string[] itemIds)
         {
-            //ToDo: Use Cache
-            var rules = Array.Empty<DynamicAssociation>();
+            var cacheKey = CacheKey.With(GetType(), nameof(GetByIdsAsync), string.Join("-", itemIds.OrderBy(x => x)));
 
-            if (!itemIds.IsNullOrEmpty())
+            var result = await _platformMemoryCache.GetOrCreateExclusiveAsync(cacheKey, async cacheEntry =>
             {
-                using (var repository = _repositoryFactory())
+                var rules = Array.Empty<DynamicAssociation>();
+
+                if (!itemIds.IsNullOrEmpty())
                 {
+                    var catalogRepository = _catalogRepositoryFactory();
                     //Optimize performance and CPU usage
-                    repository.DisableChangesTracking();
+                    catalogRepository.DisableChangesTracking();
 
-                    var entities = await repository.DynamicAssociations.Where(x => itemIds.Contains(x.Id)).ToArrayAsync();
+                    var entities = await catalogRepository.DynamicAssociations.Where(x => itemIds.Contains(x.Id)).ToArrayAsync();
 
                     rules = entities
                         .Select(x => x.ToModel(AbstractTypeFactory<DynamicAssociation>.TryCreateInstance()))
                         .ToArray();
+
+                    cacheEntry.AddExpirationToken(DynamicAssociationCacheRegion.CreateChangeToken(rules));
```

Comment:
Need to add tokens by ids to prevent caching null values for non-created instances.